Sign-in Tool Survey

Mjtlittle opened the pull request:
### Changes
- Added frontend tool survey after entering a username
   - used determine which tool the visitor plans on using in the makerspace
   - tool list specifically defined for each location
   - **Note:** currently has placeholder information for each makerspace, tool lists are not final
- Adjusted backend to accept `tool` parameter when logging a visit
   - stores tool alongside location and username in visit entry
      - tool is not a required parameter, defaults to `' '` when not provided
   - `was_user_registered` gets returned from request, indicating whether the user was registered or not; this could be used by the frontend to alert the user of the email they should receive
   - changed how error handling works in add visit entry function

### Blocked By

- [x] Finalize what tools should be at which locations
  - should the frontend be the source of truth? just like it is with majors, minors, etc.?
- [x] Needs: #85
- [x] Implementation of the backend
- [x] Planning when will it be merged

Commit:
Added tool survey after username entry

dggorsen reviewed (changes requested):
Pretty cool! How did we come up with the survey question? Is this something that will be static? Or do we envision adding variety to the questions? 

Side note- can we decouple the Sign-in tool survey PR from the pipeline oriented things?

Mjtlittle commented:
> Pretty cool! How did we come up with the survey question? Is this something that will be static? Or do we envision adding variety to the questions?

When we met with the makerspace last week, they mentioned wanting some way to get statistics on which tools are being used. Not sure if they want to collect any other information, but I re-wrote the sign-in logic to allow for further extension in the future, if we wanted to add another card after the survey or somewhere in-between.

```ts
const stages = [VisitorTypeStage, UsernameStage, ToolSurveyStage];
```

I also added the ability to change which tools are located at each location. Having an empty list skips that stage of the sign-in.

```ts
export const locations: MakerspaceLocation[] = [
  {
    slug: "watt_center",
    name: "Watt Family Innovation Center",
    tools: ["3D Printer", "Lazer Cutter", "Other"],
  },
  {
    slug: "cooper_library",
    name: "Cooper Library",
    tools: ["3D Printer"],
  },
  {
    slug: "cook_lab",
    name: "Cook Laboratory",
    tools: [],
  },
];
```

---

> Side note- can we decouple the Sign-in tool survey PR from the pipeline oriented things?

Sure, I just had it based off the branch I was working on before. I would prefer it comes after the vite env changes though, not sure if I had the `api_endpoint` var placed in the constants file before then. But I see no problem doing that.

Commit:
Implemented backend for tool survey

Mjtlittle commented:
> can we decouple the Sign-in tool survey PR from the pipeline oriented things?

Done! #85 still needs to be merged for this PR to work though

Mjtlittle commented:
I went ahead and implemented the backend aswell for this; just so we didn't have this outstanding draft PR. Only things left to do are merge #85 and figure out the exact tools the makerspace wants to show for each location.

ophilli commented:
Blocked on #85. Some additional changes incoming.

Commit:
Fixed timezone and made major/minor a string set

Commit:
Added optional choice limit to multiselect

Commit:
Updated tool lists and styling

Mjtlittle commented:
### Additional Changes

- Fixed timezone issue makerspace was observing in database
- Added ability to limit the number of selections for the major/minor registration field
- Updated the tool list from information provided from makerspace

---

**This pr should now be good to merge**

ophilli commented on cdk/visit/lambda_code/log_visit/log_visit.py:
```diff
@@ -117,12 +119,14 @@ def registrationWorkflow(self, current_user):
         # Get the current date at which the user logs in.
-        visit_date = datetime.datetime.now().timestamp()
+        eastern_tz = dateutil.tz.gettz('US/Eastern')
+        visit_date = datetime.datetime.now(tz=eastern_tz)
+        
         # Convert visit_date to human-readable format
-        visit_date = datetime.datetime.fromtimestamp(
-            visit_date).strftime('%Y-%m-%d %H:%M:%S')
+        visit_date = visit_date.strftime('%Y-%m-%d %H:%M:%S')
```
I've gotta push back on storing time in our database in a human-readable format. The unix timestamp is much easier for computers to store and manipulate than a human-readable format.

It is better to store things in your database in a way that makes sense for your application, and then convert them to human-readable at presentation time.

Mjtlittle replied:
Makes sense ill go ahead and make that change aswell

Commit:
Replaced time string with unix timestamp number

Commit:
Fixed test cases

Commit:
Fixed dates to match current prod, I think

Commit:
Reverted the string set change and added a diagram

Mjtlittle commented:
This PR should be ready to merge. The most recent two commits revert how major/minors are stored in the new tables and how the date is being stored in the old table. We aligned as a team on what this PR handles and how the migration will work.

ophilli commented on cdk/visit/lambda_code/log_visit/log_visit.py:
```diff
@@ -117,45 +119,45 @@ def registrationWorkflow(self, current_user):
-        visit_date = datetime.datetime.fromtimestamp(
-            visit_date).strftime('%Y-%m-%d %H:%M:%S')
+        eastern_tz = dateutil.tz.gettz('US/Eastern')
+        visit_date = datetime.datetime.now(tz=eastern_tz)
```
What exactly does the time zone do here? In [the docs](https://docs.python.org/3/library/datetime.html#datetime-objects) I see the following:

> classmethod datetime.now(tz=None)
> Return the current local date and time.
> 
> If optional argument tz is None or not specified, this is like [today()](https://docs.python.org/3/library/datetime.html#datetime.datetime.today), but, if possible, supplies more precision than can be gotten from going through a time.time() timestamp (for example, this may be possible on platforms supplying the C gettimeofday() function).
> 
> If tz is not None, it must be an instance of a tzinfo subclass, and the current date and time are converted to tz’s time zone.
> 
> This function is preferred over today() and [utcnow()](https://docs.python.org/3/library/datetime.html#datetime.datetime.utcnow).

Does this mean that we’d be baking the UTC-5 (or UTC-4 depending on DST) offset into our Unix time stamps with this change? A Unix time stamp is the number of seconds that have elapsed since the Unix epoch, excluding [leap seconds](https://en.m.wikipedia.org/wiki/Leap_second). The Unix epoch is 00:00:00 [UTC](https://en.m.wikipedia.org/wiki/Coordinated_Universal_Time) on 1 January 1970. 

It doesn’t really make sense to bake the timezone into a Unix time stamp, except for at the moment that you’re displaying the data in a human readable format. Does this actually change the number of seconds we save to the ddb table?

Mjtlittle replied:
This is a good point. I didn't think about any of this. I definitely agree the normalized UTC-Z time is better. I switched to use the time library in python to handle this, as it required less code to accomplish the same thing. Also left the datetime object for the original table and reverted the timezone there, so all prior data in the old table will be the same as before.

ophilli commented on cdk/visit/lambda_code/log_visit/log_visit.py:
```diff
@@ -164,65 +166,60 @@ def handle_log_visit_request(self, request, context):
+                'body': json.dumps(body)
             }
 
+        # if no request is provided (should never be the case because of gateway evocation)
```
Is evocation a real term APIGW uses? Do you mean invocation?

Mjtlittle replied:
Fixed!

ophilli commented on cdk/visit/lambda_code/register_user/register_user.py:
```diff
@@ -85,27 +87,24 @@ def add_user_info(self, user_info):
-        minorList = []
-        for minor in user_info.get('Minor', []):
-            minorList.append({'S': minor})
+        marshal_item = lambda s: {"S": s}
```
Can you add a brief comment about what `marshall_item` means? Future devs likely won’t know what marshaling is.

Mjtlittle replied:
Rewrote this part. Would have looked much nicer if I could chain the map (like in js/ts), but I feel like the functional method in python was more cluttered. Now im using a list comprehension instead, which I think is much cleaner and easier to read. Also added a comment alluding to the attribute type format.

ophilli commented on site/visitor-console/src/library/constants.ts:
```diff
@@ -7,9 +7,40 @@ export const api_endpoint =
+    slug: "watt_center",
+    name: "Watt Family Innovation Center",
+    tools: [
+      "3D Print (FDM - Plastic)",
```
`3D print` and `Resin print` should probably both be “Printer” not “Print”

Mjtlittle replied:
Revised the items to refer to the tools themselves and not any action associated with the tool; like a "print" of a "3d printer".

ophilli reviewed (commented):
This looks really good! Left you a few nitpicks comments. The only one that I think is blocking is the question I have about the Unix timestamps.

Commit:
Timezone revert and wording changes

ophilli commented on cdk/visit/lambda_code/log_visit/log_visit.py:
```diff
@@ -121,32 +120,31 @@ def registrationWorkflow(self, current_user):
-        # Get the current date at which the user logs in.
-        eastern_tz = dateutil.tz.gettz('US/Eastern')
-        visit_date = datetime.datetime.now(tz=eastern_tz)
+        timestamp = int(time.time())
```
See also https://docs.python.org/3/library/time.html#time.time_ns

ophilli commented:
@Mjtlittle I've merged in your changes. Can you test them once tomorrow after they've deployed to beta? I'd love to promote this change from beta to prod one evening this week if possible.

Mjtlittle commented:
@ophilli I just checked on the deployment, it looks like it failed to update the database in beta, as it is in use by the visitors stack. I think this is because I changed the type of the sort key for the visits table from a string to a number for storing the date. I had some issues with this in my private stack; but I didn't look into it enough to know whether it would affect the deployed stack.
